### PR TITLE
StatVar MCF and Schema MCF for UNData Energy Statistics

### PR DESCRIPTION
**Pull request by ajaits**
[Sample stat var observations from MCF previewer](https://datacommonsorg.github.io/tools/mcf-browser/#&file=https://raw.githubusercontent.com/datacommonsorg/data/b1560bef9b5bcc3e20682c1341230fec855d78af/scripts/un/energy/test_data/un_energy_output.tmcf&file=https://raw.githubusercontent.com/datacommonsorg/data/b1560bef9b5bcc3e20682c1341230fec855d78af/scripts/un/energy/test_data/un_energy_output.csv)

### CODE REVIEW

**Comment by beets** [comment]
feel free to add *.swp to this file: https://github.com/datacommonsorg/data/blob/master/.gitignore
other files that are not needed anymore can be deleted.
if there's state elsewhere (e.g. log files etc), you can use [git stash](https://www.atlassian.com/git/tutorials/saving-changes/git-stash), or save those files to another branch.

**Comment by ajaits** [comment]
Thanks Carolyn. I added *.swp to my local ~/.gitignore and ran:
git config --global core.excludesfile ~/.gitignore

The additional files have been removed now.

File: scripts/un/energy/country_codes.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""UNData Country codes mapped to DataCommons"""
+
+from typing import List
+
+# Map from UNData country codes to dcids.
+# This includes some older country coded marked as '(additional)'
```
[the remark this answers is not in the document]
Comment (ajaits) [reply scripts/un/energy/country_codes.py:19]: In the Energy dataset, they were actually older codes for countries before they were split/merged, etc. It turns out the data is replicated with the latest country code as well for such countries, so the script drops these codes now.

File: scripts/un/energy/test_data/un_energy_input.csv [hunk 2]
```diff
@@ -0,0 +1,2942 @@
+"Commodity Code","Country or Area Code","Country or Area","Transaction Code","Commodity - Transaction Code","Commodity - Transaction","Year","Unit","Quantity","Quantity Footnotes"
+"AL","32","Argentina","01","AL01","Biogasoline - production","2019","Metric tons,  thousand","850",""
```
[the remark this answers is not in the document]
Comment (ajaits) [reply scripts/un/energy/test_data/un_energy_input.csv]: This includes a line for each country code, every transaction code so all statvars are generated in the test.

The test runs is less than a second. Is it causing any issue on github?

File: scripts/un/energy/country_codes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajaits) [reply scripts/un/energy/country_codes.py:19]: The UN data set uses ISO 3166 numeric codes and they are in country/AFG nodes under property: countryNumeric3Code.
Couldn't find the schema for the property countryNumeric3Code at: https://datacommons.org/browser/countryNumeric3Code

File: scripts/un/energy/test_data/un_energy_input.csv (hunk 2 above)
[the remark this answers is not in the document]
Comment (pradh) [reply scripts/un/energy/test_data/un_energy_input.csv]: Typically, in unit-tests we only need a small subset of the data to ensure basic regression. At 21K lines of MCF the diffs don't even render on github; smaller diffs have a higher chance of manually inspecting and finding issues.

For instance, a line for every country code feels not very necessary.

File: scripts/un/energy/un_energy_codes.py [hunk 3]
```diff
@@ -0,0 +1,1431 @@
...
+        return
+    mod_lines = 1
+    if 'debug_lines' in counters and counters['debug_lines'] > 0:
+        mod_lines = counters['debug_lines']
+    if counters[counter_name] % mod_lines == 0:
+        print("Error: ", counter_name, ": ", error_msg)
+    counters[counter_name] += 1
+
+
+def remove_prefix(value: str, prefix: str) -> str:
```
Comment (pradh) [inline scripts/un/energy/un_energy_codes.py:1021]: style nit: consider making local methods prefixed with _

File: scripts/un/energy/un_energy_codes.py [hunk 4]
```diff
@@ -0,0 +1,1431 @@
...
+            code = remove_prefix(code, code)
+
+    if len(code) > 0:
+        _add_error_counter(
+            f'error_ignored_capacity_code_{code}',
+            f'Ignored capacity code: {orig_code}, suffix: {code}', counters)
+    return True
+
+
+def add_pv_for_property(code: str,
```
Comment (pradh) [inline scripts/un/energy/un_energy_codes.py:1275]: this seems complicated :)

File: scripts/un/energy/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+## Run
+
+Run all following commands from the data/scripts directory.
+
+To download and process the data:
+```bash
+python3 un/energy/process.py
+```
+This downloads multiple csv files into a tmp_data_dir/ folder and generates .csv with accompanying .mcf and .tmcf.
```
Comment (pradh) [inline scripts/un/energy/README.md:49]: nit: consider noting how many CSV files there are / how long it takes to download?

File: scripts/un/energy/process.py [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+    # Replace all '_' with a capitalized letter.
+    # Find all occurences of '_'.
+    upper_idx = [-1] + [i for i, e in enumerate(name) if e == '_'] + [len(name)]
+    # Capitalize the next letter after '_'.
+    words = [
+        name[x + 1].upper() + name[x + 2:y]
+        for x, y in zip(upper_idx, upper_idx[1:])
+        if name[x:y] != '_'
+    ]
+    return ''.join(words)
```
Comment (pradh) [inline scripts/un/energy/process.py:167]: This function and the ones above seem reusable...

**Review by pradh — APPROVED** [review]

File: scripts/un/energy/un_energy_codes.py [hunk 7]
```diff
@@ -0,0 +1,1431 @@
...
+    }
+}
+
+# Energy Loss codes
+# Prefixed with 10*
+# Types of EnergyLossEnum
+# Used as values for property energyLostAs
+UN_ENERGY_LOSS_CODES = {
+    '101': {
+        'measuredProperty': 'dcid:loss'
```
[the remark this answers is not in the document]
Comment (ajaits) [reply scripts/un/energy/un_energy_codes.py:915]: Removed 'dcid:' prefixes. I assume all properties can be referred with dcid: even if they are in the base schema with dcs:?

File: scripts/un/energy/un_energy_codes.py (hunk 3 above)
Comment (ajaits) [reply scripts/un/energy/un_energy_codes.py:1021]: Done

File: scripts/un/energy/un_energy_codes.py (hunk 4 above)
Comment (ajaits) [reply scripts/un/energy/un_energy_codes.py:1275]: Easier than defining a map with 200+ codes :)

File: scripts/un/energy/README.md (hunk 5 above)
Comment (ajaits) [reply scripts/un/energy/README.md:49]: Done

File: scripts/un/energy/process.py (hunk 6 above)
Comment (ajaits) [reply scripts/un/energy/process.py:167]: Should i move them to a common data/util, say 'counters.py' and 'dcid_utils.py'?

Q: How do I import from other parts of the data tree?
  It seems like the script needs to be run from root to allow such imports:
  from util import counters

File: scripts/un/energy/un_energy_codes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pradh) [reply scripts/un/energy/un_energy_codes.py]: Yes, correctness wise.  But for schema, generally prefer "dcs:" if we need to pick one.

File: scripts/un/energy/process.py (hunk 6 above)
Comment (pradh) [reply scripts/un/energy/process.py:162]: Feels like this the latter might have overlap with https://github.com/datacommonsorg/data/pull/489/files.

Option 1: https://github.com/datacommonsorg/data/blob/master/scripts/india_plfs/daily_wage_data/preprocess.py#L25
 (implication is we can only run this script from scripts/)
Option 2: https://github.com/datacommonsorg/data/blob/cdf4c6532be3e55dc91b621bf2b28ee533ad09ac/scripts/us_eia/eia_860/utils.py#L25-L26 (weirder but more flexible I think).

This also ties in with the discussion of structuring the py modules/tests in a consistent way in the whole repo.

File: scripts/un/energy/country_codes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajaits) [reply scripts/un/energy/country_codes.py:19]: Added a TODO to the file.

**Pull request merged**